### ISSUE TITLE
posterior_predict drawing from posterior incorrectly when link = 'log'

### ISSUE THREAD

**Issue by DanOvando** [stan-dev/rstanarm#115]
#### Summary:
posterior_predict of stan_glm object with log link drawing from posterior predictive distribution using expected value on the response scale, not link scale
#### Description:

Hi guys, loving `rstanarm` so far, thanks for making Bayesian regression quick and easily usable. I'm running a glm with a log link function (`family = gaussian(link = 'log')`), and trying to get draws from posterior predictive using the `posterior_predict` function. However, it seems that posterior_predict is taking draws from the gaussian distribution with a mean on the response scale, not on the link scale. So, as an example, rather than taking draws from `rnorm(1, log(mu), sigma)`, as we would want when using a glm with `family = gaussian(link = 'log')` with dependent variable mu , `posterior_predict` appears to be taking draws from `rnorm(1, mu, sigma)`. 

Hard to notice except when mu is small and/or sigma is large enough, posterior_predict produces negative draws of mu on the response scale, which should be impossible, given that we expect log(mu) to be normally distributed,  not mu (which should be log-normal as modeled, and therefore >0)

https://github.com/stan-dev/rstanarm/blob/master/R/posterior_predict.R#L190-L194

#### Reproducible Steps:

Example using mtcars, in the manner of `help('posterior_predict')`

```
library(rstanarm)
library(tidyverse)

mtcars2 <- mtcars %>%
  mutate(mpg = mpg/100) #get things close to zero so the problem pops up

fit_loglink_glm <- stan_glm(mpg ~ wt, data = mtcars2, family = gaussian(link = 'log'))

ytilde <- posterior_predict(fit_loglink_glm, seed = 123) %>%
  as_data_frame() %>%
  gather('obs','i') %>%
  mutate(obs = str_replace(obs,'^.','') %>% as.numeric()) %>%
  group_by(obs) %>%
  summarise(ytilde = mean(i),
            yupper = quantile(i,.975),
            ylower = quantile(i, 0.025),
            min_ytilde = min(i)
            ) %>%
  ungroup() %>%
  mutate(y_true = as.numeric(mtcars2$mpg))

#Evidence that posterior predict is producing draws on the response scale,
#but from a gaussian distribution, producing negative draws of ytilde

ytilde %>%
  ggplot() +
  geom_pointrange(aes(y_true, ytilde, ymax = yupper, ymin = ylower)) +
  geom_abline(aes(intercept = 0, slope = 1 )) +
  geom_hline(aes(yintercept = 0))

```
#### RStanARM Version:
The version of the rstanarm package you are running (e.g., from `packageVersion("rstanarm")`)
2.11.1

#### R Version:
The version of R you are running (e.g., from `getRversion()`)

‘3.3.1’

#### Operating System:
Your operating system (e.g., OS X 10.11.3)

OS X 10.11.6

**Comment by jgabry** [comment]
Hi Dan, glad you like rstanarm! And thanks for opening the issue. I don't think what you're reporting here is actually bug, but it did lead us to identify a small bug that's related to this. 

First, a lognormal model and a gaussian GLM with log link are different models. The former uses a lognormal as the conditional PDF for the outcome `y` whereas the latter uses the normal distribution with an exponentiated linear predictor. So `family = gaussian(link="log")` will guarantee that the conditional mean is positive (i.e. `mu = exp(a + bx)` will be positive), but this does not guarantee that simulated _outcomes_ are all positive because it's possible to draw from a normal distribution with a positive mean and still get negative numbers. Fitted values (in sample point predictions) for y would be positive (like they are when you use plain old `glm`), but not necessarily every individual draw from the posterior predictive distribution. 

So when you say

> Hard to notice except when mu is small and/or sigma is large enough, posterior_predict produces negative draws of mu on the response scale, which should be impossible, given that we expect log(mu) to be normally distributed, not mu (which should be log-normal as modeled, and therefore >0)
https://github.com/stan-dev/rstanarm/blob/master/R/posterior_predict.R#L190-L194

you're kind of right, in that you will get a few negative draws if `mu` is small and/or `sigma` is large, but that's because `mu` being positive and the predicted `y` being positive isn't the same thing. Also, the `mu` that's used in the lines of code you reference above has already been transformed by the inverse link function (in this case `exp`) elsewhere in the code: 
https://github.com/stan-dev/rstanarm/blob/master/R/posterior_predict.R#L269

A lognormal data model, on the other hand, would always give you positive predictions for the outcome because it assumes that y itself is always positive rather than just E(y|•). The way to implement the lognormal model using rstanarm would be to log-transform the outcome variable and then use `gaussian(link = "identity")` rather than `link="log"`. That is to say that log(y) is conditionally normal (not log(mu)). Then the predictions would be for `log(y)` and exponentiating them would give you predictions for `y` (`posterior_predict` accepts a `fun` argument for this purpose).

Having said all that, we did find a small bug based on your issue, just not exactly what you're describing. We were actually using the lognormal likelihood rather than just the normal with exponentiated linear predictor. We'll fix that, but that's related to the code for the likelihood not anything in `posterior_predict` and fixing it won't prevent you from getting a few negative predictions if you do actually use the log link instead of log-transforming y. 

Hopefully that all makes sense. I'm not sure if I expressed all that very clearly so definitely feel free to follow up it that's not clear or if you disagree!

Jonah
